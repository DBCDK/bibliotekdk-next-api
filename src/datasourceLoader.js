import DataLoader from "dataloader";
import { log } from "dbc-node-logger";
import { withRedis } from "./datasources/redis.datasource";
import monitor from "./utils/monitor";
import { getFilesRecursive } from "./utils/utils";

// Find all datasources in src/datasources
export const datasources = getFilesRecursive("./src/datasources")
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status } = require(file.path);
    if (!load) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    // Monitor the load function from the datasource
    const monitoredLoad = monitor(
      {
        name: `REQUEST_${name}`,
        help: `${name} requests`,
      },
      load
    );

    // if datasource exports a createBatchLoader we use that,
    // otherwise we use default batch loader
    let monitoredBatchLoader = batchLoader
      ? (keys) => batchLoader(keys, monitoredLoad)
      : async (keys) => {
          return await Promise.all(keys.map((key) => monitoredLoad(key)));
        };

    // Check if Redis is configured for this datasource
    if (options && options.redis && options.redis.prefix && options.redis.ttl) {
      monitoredBatchLoader = withRedis(monitoredBatchLoader, {
        prefix: options.redis.prefix,
        ttl: options.redis.ttl,
      });
    }

    const statusChecker = status && (async () => await status(monitoredLoad));

    return {
      batchLoader: monitoredBatchLoader,
      name,
      options,
      statusChecker,
    };
  })
  .filter((func) => !!func);

log.info(`found ${datasources.length} datasources`, {
  datasources: datasources.map((datasource) => ({
    name: datasource.name,
    options: datasource.options,
  })),
});

/**
 * Will instantiate dataloaders from datasources
 */
export default function createDataLoaders() {
  const result = {};
  datasources.forEach((datasource) => {
    result[datasource.name] = new DataLoader(datasource.batchLoader, {
      // If key is an object, we stringify
      // to make it useful as a cache key
      cacheKeyFn: (key) =>
        typeof key === "object" ? JSON.stringify(key) : key,
    });
  });
  return result;
}
