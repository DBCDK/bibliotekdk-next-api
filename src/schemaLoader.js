/**
 * @file Responsible for loading the schema
 *
 * It will check all files recursively in ./src/schema
 * and load type definitions and resolvers
 */

import { makeExecutableSchema, mergeSchemas } from "graphql-tools";
import drupalSchema from "./schema/external/drupal";
import { log } from "dbc-node-logger";
import { getFilesRecursive } from "./utils/utils";

/**
 * Will check all files in schema folder
 * for type definitions and resolvers.
 *
 * These are then loaded
 */
function schemaLoader() {
  let allTypeDefs = [];
  let allResolvers = {};

  // Load files in schema folder
  const files = getFilesRecursive("./src/schema");

  // Require typeDefs and resolvers
  files.forEach((file) => {
    if (!file.path.endsWith(".js")) {
      return;
    }
    const { typeDef, resolvers } = require(file.path);
    if (typeDef) {
      allTypeDefs = [...allTypeDefs, typeDef];
      log.info(`Found type definition in ${file.path}`);
    }
    if (resolvers) {
      allResolvers = { ...allResolvers, ...resolvers };
    }
  });

  return { typeDefs: allTypeDefs, resolvers: allResolvers };
}

const { typeDefs, resolvers } = schemaLoader();

/**
 * Create executable schema from type definitions and resolvers
 */
export const internalSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

/**
 * We stitch the internal schema together with the external drupal schema
 */
export default async () => {
  return mergeSchemas({
    subschemas: [{ schema: await drupalSchema() }, { schema: internalSchema }],
  });
};
