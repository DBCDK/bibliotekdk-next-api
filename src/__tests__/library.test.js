/**
 * @file This file contains tests for the GraphQL API
 *
 * We want to ensure:
 *  - That fields can be retrieved as expected. If a type is changed, these tests should
 *    generally still work. Changes should be backwards compatible.
 *  - Error handling works.
 *  - Proper handling of insane/complex queries. For instance, highly nested queries
 *    that results in explosion of requests to underlying services.
 *
 * Error handling:
 *  - User errors. Query syntax errors, Queries that do not adhere to the schema, missing variables etc.
 *  - Datasource errors. Data from datasource with wrong format. Data not found.
 *    If we don't validate, these will probably lead to resolver errors.
 *  - Resolver errors. Bugs in resolvers, datasource errors may lead to resolver errors.
 *    We should have these logged, to be able to identify whether its a resolver bug or datasource bug
 *  - Network errors.
 *
 * It should be possible for users of the API to discern from the respone what type of error
 * they are getting. THIS IS STILL WORK IN PROGRESS
 *
 */

import { graphql } from "graphql";
import { internalSchema } from "../schemaLoader";
import { createMockedDataLoaders } from "../datasourceLoader";

export async function performTestQuery({ query, variables, context }) {
  return graphql(internalSchema, query, null, context, variables);
}

test("library - get branches", async () => {
  const result = await performTestQuery({
    query: `
          query{
            library(agencyid: "710100"){    
              branches{
                agencyId
                branchId
                name
                openingHours
              }
            }
          }
        `,
    variables: {},
    context: { datasources: createMockedDataLoaders() },
  });
  expect(result).toMatchSnapshot();
});
