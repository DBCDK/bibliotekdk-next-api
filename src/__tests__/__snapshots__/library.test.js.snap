// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`library - get branches 1`] = `
Object {
  "data": Object {
    "library": Object {
      "branches": Array [
        Object {
          "agencyId": "710100",
          "branchId": "710100",
          "city": "København K",
          "name": "Hovedbiblioteket, Krystalgade",
          "openingHours": "Åbningstider
Se åbningstider på:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          "pickupAllowed": true,
          "postalAddress": "Krystalgade 15",
          "postalCode": "1172",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710104",
          "city": "København N",
          "name": "Blågårdens Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Blågårds Plads 5",
          "postalCode": "2200",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710105",
          "city": "Brønshøj",
          "name": "Brønshøj Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Krabbesholmvej 3
",
          "postalCode": "2700",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710106",
          "city": "København K",
          "name": "Christianshavns Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Dronningensgade 53",
          "postalCode": "1420",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710107",
          "city": "Brønshøj",
          "name": "Husum Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Frederikssundsvej 290",
          "postalCode": "2700",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710108",
          "city": "København S",
          "name": "Islands Brygge Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Njalsgade 15",
          "postalCode": "2300",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710109",
          "city": "København Ø",
          "name": "Øbro Jagtvej Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Jagtvej 227",
          "postalCode": "2100",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710110",
          "city": "København NV",
          "name": "Biblioteket Rentemestervej",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Rentemestervej 76",
          "postalCode": "2400",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710111",
          "city": "København N",
          "name": "Nørrebro Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Nørrebrohallen
Nørrebrogade 208",
          "postalCode": "2200",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710112",
          "city": "København S",
          "name": "Solvang Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Remisevej 14",
          "postalCode": "2300",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710113",
          "city": "København S",
          "name": "Sundby Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Kvarterhuset Jemtelandsgade 3",
          "postalCode": "2300",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710114",
          "city": "København S",
          "name": "Bibliotekshuset",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Rodosvej 4",
          "postalCode": "2300",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710115",
          "city": "København SV",
          "name": "Sydhavnens Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Wagnersvej 19",
          "postalCode": "2450",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710116",
          "city": "Brønshøj",
          "name": "Tingbjerg Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Skolesiden 4",
          "postalCode": "2700",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710117",
          "city": "Valby",
          "name": "Valby Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Annexstræde 2",
          "postalCode": "2500",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710118",
          "city": "Vanløse",
          "name": "Kulturstationen Vanløse. Biblioteket",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Jernbane Alle 38",
          "postalCode": "2720",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710119",
          "city": "København V",
          "name": "Vesterbro Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Lyrskovgade 4",
          "postalCode": "1758",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710120",
          "city": "Valby",
          "name": "Vigerslev Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Kirsebærhaven 23",
          "postalCode": "2500",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710121",
          "city": "København Ø",
          "name": "Østerbro Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Dag Hammarskjölds Alle 19",
          "postalCode": "2100",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710122",
          "city": "København S",
          "name": "Ørestad Bibliotek",
          "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          "pickupAllowed": true,
          "postalAddress": "Arne Jacobsens Allé 19",
          "postalCode": "2300",
        },
        Object {
          "agencyId": "710100",
          "branchId": "710123",
          "city": "Albertslund",
          "name": "Københavns Bibliotekers centrale sortering",
          "openingHours": "ma-fre: 9-16",
          "pickupAllowed": true,
          "postalAddress": "Djursvang 7C",
          "postalCode": "2620",
        },
      ],
    },
  },
}
`;
